packages/codemirror/src/searchprovider.ts

```diff
@@ -30,521 +30,362 @@
-    const caseSensitive = this._query.ignoreCase;
-    const state = this.editor.state;
-
-    // In order to support search-as-you-type, we needed a way to allow the first
-    // match to be selected when a search is started, but prevent the selected
-    // search to move for each new keypress.  To do this, when a search is ended,
-    // the cursor is reversed, putting the head at the 'from' position.  When a new
-    // search is started, the cursor we want is at the 'from' position, so that the same
-    // match is selected when the next key is entered (if it is still a match).
-    //
-    // When toggling through a search normally, the cursor is always set in the forward
-    // direction, so head is always at the 'to' position.  That way, if reverse = false,
-    // the search proceeds from the 'to' position during normal toggling.  If reverse = true,
-    // the search always proceeds from the 'anchor' position, which is at the 'from'.
-
-    const lastPosition = reverse
-      ? state.selection.main.anchor
-      : state.selection.main.head;
-    let cursor = new RegExpCursor(
-      this.editor.doc,
-      this._query.source,
-      { ignoreCase: !caseSensitive },
-      lastPosition
-    ).next();
-    if (cursor.done) {
-      const startOrEnd = reverse ? this.editor.doc.length : 0;
-      cursor = new RegExpCursor(
-        this.editor.doc,
-        this._query.source,
-        { ignoreCase: !caseSensitive },
-        startOrEnd
-      ).next();
-      if (cursor.done) {
-        return null;
-      }
+  protected async onSharedModelChanged(
+    emitter: ISharedText,
+    changes: SourceChange
+  ): Promise<void> {
+    if (changes.sourceChange) {
+      await this.updateCodeMirror(emitter.getSource());
+      this._stateChanged.emit();
     }
-    return {
-      from: this._getPosition(cursor.value.from),
-      to: this._getPosition(cursor.value.to)
-    };
-  }
-
-  // TODO: remove this when Position API are removed and we work with
-  // offsets only. Contrary to the functions in CodeMirrorEditor,
-  // these functions work with CM6 position (i.e. line number starting at 1).
-  private _getPosition(offset: number): { line: number; ch: number } {
-    const line = this.editor.doc.lineAt(offset);
-    return { line: line.number, ch: offset - line.from };
   }
 
-  private _parseMatchesFromState(): ISearchMatch[] {
-    // Flatten state map
-    const matches = new Array<ISearchMatch>();
-
-    for (const lineKey in this._matchState) {
-      const lineMatches = this._matchState[lineKey];
-      for (const posKey in lineMatches) {
-        matches.push(lineMatches[posKey]);
+  protected async updateCodeMirror(content: string) {
+    if (this.query !== null) {
+      if (this.isActive) {
+        this.cmHandler.matches = await TextSearchEngine.search(
+          this.query,
+          content
+        );
+      } else {
+        this.cmHandler.matches = [];
```

Comment:
Should we always fallback to `[]` even when `query` is null? To be sure to clean any search?